client_cli/src/main.rs

```diff
@@ -117,21 +117,29 @@ pub trait RunArgs {
     ///
     /// # Errors
     /// if inner command errors
-    fn run(self, cfg: &ClientConfiguration) -> Result<Box<dyn Serialize>>;
+    fn run<F: FnMut(&dyn Serialize) -> Result<()>>(
+        self,
+        cfg: &ClientConfiguration,
+        callback: F,
+    ) -> Result<()>;
```

Comment:
I don't like that each `RunArgs` implementor has to accept a magic `callback` and pass it explicitly to the final destination. Also, each time you have to specify the type of callback, which is verbose.

I think it would be better to introduce a trait like `RunContext`:

```rust
trait RunContext {
  fn configuration(&self) -> &ClientConfiguration;

  fn print_data(&self, value: impl Serialize) -> Result<()>;
}
```

That is something you can pass everywhere instead of `cfg` + `callback`. This approach is even less cost in terms of runtime (you don't need to pass 2 args, and you can simply `impl RunContext` with callback instead of passing a closure).

Reply:
Runtime cost is definitely not an issue, of course. But I wonder if code organisation will be better with an umbrella trait.

Reply:
I will try to implement your approach. I think it will be look nicer that current one.